### PR TITLE
drivers: Add Atmel SAM counter (TC) Driver

### PR DESCRIPTION
**Pull request by mnkp**
Add basic counter driver based on Timer Counter (TC) module for Atmel SAM family.

Remarks:
- The driver is not thread safe.
- The driver does not implement guard periods.
- The driver does not guarantee that short relative alarm will trigger the interrupt immediately and not after the full cycle / counter overflow.

Use at your own risk or submit a patch.

Tested on Atmel SMART SAM E70 Xplained board

Depends on #zephyrproject-rtos/hal_atmel/pull/15

### CODE REVIEW

File: drivers/counter/counter_sam_tc.c [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+	32768,
+	USEC_PER_SEC, USEC_PER_SEC, USEC_PER_SEC,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ,
+};
+
+static int counter_sam_tc_start(const struct device *dev)
+{
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
```
Comment (stephanosio) [inline drivers/counter/counter_sam_tc.c:82]: AFAICS, `TcChannel` is not available in other SAM series headers (i.e. `sam3*` and `sam4*`).

File: drivers/counter/counter_sam_tc.c [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+	.get_pending_int = counter_sam_tc_get_pending_int,
+};
+
+#define COUNTER_SAM_TC_INPUT_FREQUENCY(n)	\
+		sam_tc_input_freq_table[DT_INST_PROP(n, clk##_ENUM_IDX)]
+
+#define COUNTER_SAM_TC_CMR(n)	\
+		 (TC_CMR_TCCLKS(DT_INST_PROP(n, clk##_ENUM_IDX)) \
+		| TC_CMR_WAVEFORM_WAVSEL_UP_RC \
+		| TC_CMR_WAVE)
```
Comment (stephanosio) [inline drivers/counter/counter_sam_tc.c:331]: Same deal here. These definitions are only available in the `sam7x` headers.

File: drivers/sensor/qdec_sam/qdec_sam.c [hunk 3]
```diff
@@ -45,7 +45,7 @@ static int qdec_sam_fetch(const struct device *dev, enum sensor_channel chan)
 	const struct qdec_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
 	struct qdec_sam_dev_data *const dev_data = DEV_DATA(dev);
 	Tc *const tc = dev_cfg->regs;
-	TcChannel *tc_ch0 = &tc->TC_CHANNEL[0];
+	TcChannel *tc_ch0 = &tc->TcChannel[0];
```
Comment (stephanosio) [inline drivers/sensor/qdec_sam/qdec_sam.c:48]: Same deal here.

**Review by stephanosio — COMMENTED** [review]
It looks like this won't work on other SAM family devices (i.e. `sam3x` and `sam4x` series). We should be patching the rest of the SAM family device headers in `hal_atmel` as well.

p.s. alternatively, https://github.com/zephyrproject-rtos/zephyr/pull/36181#discussion_r650346165

File: dts/bindings/timer/atmel,sam-tc.yaml [hunk 4]
```diff
@@ -21,6 +21,41 @@ properties:
...
+
+    clk:
+      type: string
+      required: true
+      enum:
+        - "PCK6"
+        - "MCK_8"
+        - "MCK_32"
+        - "MCK_128"
+        - "SLCK"
```
Comment (nandojve) [inline dts/bindings/timer/atmel,sam-tc.yaml:39]: Clocks don't align with all SoC:
SAM4L uses different  Generic Clock instances at TIMER_CLOCK1 for each TC instance. The clock uses BUS clock and divisors are different.
SAM3X/4S/4E uses MCK_2 at TIMER_CLOCK1
SAME7x/V7x uses PCK6 at TIMER_CLOCK1

My suggestion is use TCK_1, TCK_2, .. TCK_5 instead. With that, we can resolve in code and have only one driver.

File: drivers/counter/counter_sam_tc.c [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+
+static const uint32_t sam_tc_input_freq_table[] = {
+	USEC_PER_SEC,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 8,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 32,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 128,
+	32768,
+	USEC_PER_SEC, USEC_PER_SEC, USEC_PER_SEC,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ,
+};
```
Comment (nandojve) [inline drivers/counter/counter_sam_tc.c:76]: Clocks don't align with all SoC:
SAM3X/4S/4E uses MCK_2 at TIMER_CLOCK1
SAME7x/V7x uses PCK6 at TIMER_CLOCK1

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @mnkp ,

I agreed with @stephanosio "_We should be patching the rest of the SAM family device headers in hal_atmel as well_".

I looked at SAM TC and try bring all differences:

SAM3X/SAM4L/SAM4S doesn't have `0x0C Register AB`. It is usually read by the DMA and that explains the why.
SAM3X don't implement _2-bit gray up/down count for stepper motor_

SAM4L This is a low power SoC. The implementation is reduced and don't have:
```
0x30 Extended Mode Register
0xC0 Block Control Register
0xC4 Block Mode Register
0xC8 QDEC Interrupt Enable Register
0xCC QDEC Interrupt Disable Register
0xD0 QDEC Interrupt Mask Register
0xD4 QDEC Interrupt Status Register
0xD8 Fault Mode Register
0xE4 Write Protection Mode Register
```

DMA is implemented only on SAM4E by PDC. In this case, `0x100–0x1A4 Reserved for PDC Registers`.
SAME7x/V7x uses XDMAC.

SAM4L should be excluded at Kconfig right now to keep it simple change hal_atmel.

**Review by mnkp — COMMENTED** [review]
I was hoping to avoid providing support for all Atmel SAM family and limit it only to SAM E70/V71 series for now. I have only SAM E70 Xplained board. @stephanosio @nandojve if any of you would be willing to test the driver on SAM3X/4S/4E SoCs I will update the PR.

@nandojve thanks for the detailed analysis. From all the differences that were mentioned the one that affects the counter driver is `0x30 Extended Mode Register` since I'm using it to optionally drive the TC module directly from the MCLK. This is not possible via the standard method using TCCLKS: Clock Selection bit-field. I will change the .yaml bindings file to include MCLK as a separate property with appropriate documenation and then add some #ifdef in the driver.

> SAM4L should be excluded at Kconfig right now to keep it simple change hal_atmel.
  
It's possible to enable the driver only on SoC series that define tc? nodes in the DTS file. Since they contain `compatible = "atmel,sam-tc"` property. At present only SAM E70, V71 series do.

File: dts/bindings/timer/atmel,sam-tc.yaml (hunk 4 above)
Comment (nandojve) [reply dts/bindings/timer/atmel,sam-tc.yaml:39]: It is up to you, I don't have preference. When I comment about TCK_1 .. I mean a shortcut for TIMER_CLOCK1.

**Comment by nandojve** [comment]
> I was hoping to avoid proving support for all Atmel SAM family and limit it only to SAM E70/V71 series for now. I have only SAM E70 Xplained board. @stephanosio @nandojve if any of you would be willing to test the driver on SAM3X/4S/4E SoCs I will update the PR.

I have 4L/4S/4E/V71 boards here. If you decide move forward and update hal_atmel for anyone of those I can run TC tests to help you.

**Comment by mnkp** [comment]
I've pushed a new version that supports all SAM series. For know that HAL fixes are done as patch. Once we have a basic patch infrastructure in soc/ introduced by #36181 some (but not all of the HAL patches) can be moved there.

**Comment by nandojve** [comment]
Hi @mnkp ,

I tested with success on SAMV71 and SAM4S. For some reason my SAM4E don't trigger ISR. I need look with time to understand what could be the problem. I think it can be my board. BTW, I discovered that SAM4E TC is 32 bits. SAM4L seems to work but clock is to fast and timer couldn't be configured. I need time to enable a generic clock to enable a slow clock.

**Comment by mnkp** [comment]
@nandojve Thanks for all the help!

> For some reason my SAM4E don't trigger ISR. 

I double checked DTS to make sure I haven't made a mistake, but all looks well.

> I discovered that SAM4E TC is 32 bits.

Thanks, I missed this. I'll need to update the PR.

> SAM4L seems to work but clock is to fast and timer couldn't be configured. I need time to enable a generic clock to enable a slow clock.

Yes, unfortunately. Without a slow clock and 16-bit channel the supported time span is very short. In the counter sample one could manually shorten the time when the alarm is triggered and verify that the callback is executed. Temporarily, just to test the driver.

**Comment by nandojve** [comment]
> > SAM4L seems to work but clock is to fast and timer couldn't be configured. I need time to enable a generic clock to enable a slow clock.
> 
> Yes, unfortunately. Without a slow clock and 16-bit channel the supported time span is very short. In the counter sample one could manually shorten the time when the alarm is triggered and verify that the callback is executed. Temporarily, just to test the driver.

Added #36406 to enable a slow clock at Generic Clock 5 (TC0). This is something that allow true test with HW.

File: boards/arm/sam4l_ek/sam4l_ek.dts [hunk 6]
```diff
@@ -62,6 +62,11 @@
 		   <&gpioc 0 GPIO_ACTIVE_LOW>;
 };
 
+&tc0 {
+	clk = <4>;
```
Comment (nandojve) [inline boards/arm/sam4l_ek/sam4l_ek.dts:66]: It needs to be `clk = <0>` to select the slow clock at GCLK-5.

File: drivers/counter/counter_sam_tc.c [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+
+static const uint32_t sam_tc_input_freq_table[] = {
+#if defined(CONFIG_SOC_SERIES_SAME70) || defined(CONFIG_SOC_SERIES_SAMV71)
+	USEC_PER_SEC,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 8,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 32,
+	SOC_ATMEL_SAM_MCK_FREQ_HZ / 128,
+	32768,
+#elif defined(CONFIG_SOC_SERIES_SAM4L)
+	USEC_PER_SEC,
```
Comment (nandojve) [inline drivers/counter/counter_sam_tc.c:77]: The index 0 for SAM4L should be adjusted to 1024.
This value is generated from RC32K -> GCLK5 / 32 -> 1024 Hz.

**Review by nandojve — CHANGES_REQUESTED** [review]
Changes to enable a functional SAM4L version.

**Comment by nandojve** [comment]
Hi @mnkp ,

I conducted more tests with SAM4E. I changed source clock to `clk = <3>` and cheated at `sam_tc_input_freq_table` to force a run. The TC is working OK for SAM4E. However, SLCK don't work as clock source and I'm pretty convinced that there is nothing to do with TC driver implementation.

Please, update remaining details to move forward.

File: drivers/counter/counter_sam_tc.c [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+{
+	struct counter_sam_dev_data *data = DEV_DATA(dev);
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+	uint32_t alarm_value;
+
+	__ASSERT_NO_MSG(alarm_cfg->callback != NULL);
+
+	if (chan_id >= MAX_ALARMS_PER_TC_CHANNEL) {
```
Comment (nordic-krch) [inline drivers/counter/counter_sam_tc.c:137]: this check is already done in the api wrapper in `counter.h`

File: drivers/counter/counter_sam_tc.c [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+}
+
+static int counter_sam_tc_cancel_alarm(const struct device *dev, uint8_t chan_id)
+{
+	struct counter_sam_dev_data *data = DEV_DATA(dev);
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+
+	if (chan_id >= MAX_ALARMS_PER_TC_CHANNEL) {
```
Comment (nordic-krch) [inline drivers/counter/counter_sam_tc.c:190]: this check is already done in the api wrapper in counter.h

File: drivers/counter/counter_sam_tc.c [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+
+	/* Disable the compare interrupt */
+	tc_ch->TC_IDR = TC_IDR_CPCS;
+
+	data->top_cb = top_cfg->callback;
+	data->top_user_data = top_cfg->user_data;
+
+	tc_ch->TC_RC = top_cfg->ticks;
+
+	if ((top_cfg->flags & COUNTER_TOP_CFG_DONT_RESET) &&
```
Comment (nordic-krch) [inline drivers/counter/counter_sam_tc.c:233]: i have no idea how timer works but seems to me that there is the same behavior when  `DONT_RESET` option is set and current tick is smaller than new value and case when `DONT_RESET` option is off. Both lands in `c_ch->TC_CCR = TC_CCR_SWTRG`.

File: drivers/counter/counter_sam_tc.c [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+	struct counter_sam_dev_data *data = DEV_DATA(dev);
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+	uint32_t status;
+
+	status = tc_ch->TC_SR;
+
+	if (status & TC_SR_CPAS) {
+		if (data->alarm[0].callback) {
```
Comment (nordic-krch) [inline drivers/counter/counter_sam_tc.c:278]: shouldn't you somehow disable given interrupt so it does not get triggered on the next cycle?

File: drivers/counter/counter_sam_tc.c [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+		tc_ch->TC_IDR = TC_IDR_CPAS;
+	} else {
+		tc_ch->TC_IDR = TC_IDR_CPBS;
+	}
+
+	data->alarm[chan_id].callback = alarm_cfg->callback;
+	data->alarm[chan_id].user_data = alarm_cfg->user_data;
+
+	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) != 0) {
+		alarm_value = alarm_cfg->ticks;
```
Comment (nordic-krch) [inline drivers/counter/counter_sam_tc.c:161]: should return error when `COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE` flag is set.

File: drivers/counter/counter_sam_tc.c [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+	} else {
+		alarm_value = (tc_ch->TC_CV + alarm_cfg->ticks) % top_value;
```
Comment (nordic-krch) [inline drivers/counter/counter_sam_tc.c:163]: there is no protection against ticks being to small (e.g. 1).

It mentions above that guard periods are not supported but they apply to absolute alarms. There is an assumption that relative alarm will always expire as expected and not after full cycle.

**Comment by carlescufi** [comment]
@mnkp https://github.com/zephyrproject-rtos/hal_atmel/pull/15 is merged now.

File: drivers/counter/counter_sam_tc.c (hunk 12 above)
Comment (mnkp) [reply drivers/counter/counter_sam_tc.c:172]: I believe this is only applicable once support for guard periods is implemented.

File: drivers/counter/counter_sam_tc.c (hunk 13 above)
Comment (mnkp) [reply drivers/counter/counter_sam_tc.c]: @nordic-krch Thanks for the review! All issues should be fixed now apart from short relative alarms. Implementing this correctly is not trivial. For now I've added the follwing comment to the driver source file as well as the commit

> - The driver does not guarantee that short relative alarm will trigger the interrupt immediately and not after the full cycle / counter overflow.

I would prefer if the next person that needs a feature implements and tests it. Hope that's OK.

**Comment by mnkp** [comment]
All issues should be fixed now apart from

- support for short relative alarms
- slow clock on SAM4L, depends on #36406

The driver supports now 32-bit TC counter on SAM4E series, though the feature was not tested.

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @mnkp ,

Everything is working fine!

Just a remind, not a block, boards with revision B won't work without overlay. In case someone look this in future. `samples/drivers/counter/alarm/boards/sam_v71b_xult.overlay`

It'll be nice if you update commit messages, now TC support all SAM SoC.

File: boards/arm/sam4l_ek/sam4l_ek.dts [hunk 14]
```diff
@@ -62,6 +62,11 @@
 		   <&gpioc 0 GPIO_ACTIVE_LOW>;
 };
 
+&tc0 {
```
[the remark this answers is not in the document]
Comment (mnkp) [reply boards/arm/sam4l_ek/sam4l_ek.dts:65]: It looks inconsistent, some SoC series have overlays and other enable tc0 in the board file. The reason for this is support for QDEC mode / driver. SAM4L does not support QDEC so the counter is enabled directly in the board file. All other SoC series support QDEC. In that case I can't enable tc0 in the board file as qdec_sam driver would not compile.

**Comment by mnkp** [comment]
@nandojve thanks for the comments. Everything should be fixed now.

**Review by nandojve — APPROVED** [review]
LGTM
Thank you @mnkp !

**Review by cfriedt — APPROVED** [review]

**Pull request merged**